### PR TITLE
[YW-308] feat(dashboard): fan-out batch command — clone viz item N times with field pin

### PR DESCRIPTION
**Pull request by youhaowei**
## Summary

- Adds `FanOutDashboardItems` to the `applyCommand` vocabulary — a batch command that, given `{ dashboardId, sourceItemId, field, placements[] }`, clones a source viz item N times, each with `overrides.filters` pinning the field to the placement's value
- Source item and insight definition are never mutated — only `DashboardItem.overrides` is written, flowing through the existing draft/preview/publish pipeline
- No UI, no cardinality guard, no layout-rule logic — primitive only; the agent supplies all inputs deterministically

## Files changed

- `apps/server/src/functions/commands.ts` — `DashboardItemOverrides` interface, extend `DashboardItem`, `fanOutDashboardItems` mutation handler, registry/paths/payloads wired
- `apps/server/src/functions/preview-diff.ts` — `fanOutDashboardItemsCmd` added to exhaustive `COMMAND_DESCRIPTORS` and `PATH_TO_NAME` maps
- `packages/assistant/src/apply-command-tool.ts` — `FanOutDashboardItems` added to `DRAFT_SAFE_COMMANDS`
- `packages/assistant/src/read/command-guide.ts` — guide entry added (freshness gate passes)
- `apps/server/src/functions/commands.test.ts` — 11 new tests: N-item emit, eq-operator pin, dimension inheritance, override preservation, same-field replace, insight immutability, and all error guards

## Test plan

- [x] `command-guide-freshness.test.ts` — 3/3 (guide ⟷ registry in sync)
- [x] `commands.test.ts` — 117/117 (all existing + 11 new FanOutDashboardItems)
- [x] Full graph typecheck (`turbo typecheck`) — 48/48
- [x] `eslint` — clean on both `@dashframe/server` and `@dashframe/assistant`

Tracked internally as YW-308

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01EeGZooqvrSHBxaTvTWkjJn

<details open><summary><h3>Greptile Summary</h3></summary>

Adds `FanOutDashboardItems` — a batch command that clones a source viz panel N times, each pinning a field value in `DashboardItem.overrides`, and wires it through the full command vocabulary (registry, preview-diff, agent guide, draft-safe set). The source item and insight are never mutated; only the dashboard layout is written.

- `commands.ts` introduces `DashboardItemOverrides`/`DashboardItemFilterOverride` interfaces (with narrow operator union), the `fanOutDashboardItems` mutation handler with guards for empty placements, duplicate ids, missing source, and absent `value` key, and passes `overrides` through `requireDashboardItem`/`sanitizeDashboardItemUpdates` as opaque jsonb.
- `commands.test.ts` adds 11 focused tests covering clone count, operator pinning, dimension inheritance, override preservation, same-field replacement, insight immutability, and all error guards (including the null-vs-absent `value` edge case).
- `preview-diff.ts` and `apply-command-tool.ts` wire the new command into the exhaustive descriptor/name maps and the draft-safe set.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the fan-out primitive is well-contained, the handler only writes to `dashboards.layout`, the source item and insight definition are never touched, and all validation guards are exercised by the 11 new tests.

The mutation follows the established read-modify-write pattern for dashboard layout updates, input validation is thorough (placement shape, absent `value` key, duplicate ids, markdown source guard), and the operator union on the internal interface is now correctly narrowed. All issues from previous review threads appear to have been addressed in this version.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/server/src/functions/commands.ts | Core fan-out mutation handler added with solid validation guards; `overrides` is intentionally passed through opaque in `sanitizeDashboardItemUpdates`/`requireDashboardItem`, and `sourceSorts` spread (noted in prior thread) partially isolates from source but all N clones share the same spread array. |
| apps/server/src/functions/commands.test.ts | 11 new tests covering all key contracts: N-item emit, eq-operator, dimension inheritance, override preservation, same-field replacement, insight immutability, and all error paths including the null-vs-absent value edge case. |
| apps/server/src/functions/preview-diff.ts | Adds `fanOutDashboardItemsCmd` to the exhaustive `COMMAND_DESCRIPTORS` and `PATH_TO_NAME` maps with correct `kind`, `targetId`, and a safe summary lambda. |
| packages/assistant/src/apply-command-tool.ts | Single-line addition of `FanOutDashboardItems` to `DRAFT_SAFE_COMMANDS`; no concerns. |
| packages/assistant/src/read/command-guide.ts | New guide entry accurately describes the command contract, placement shape, dimension defaulting, and immutability guarantee; also updates `AddDashboardItem` notes to reference `FanOutDashboardItems`. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Agent
    participant applyCommand
    participant fanOutDashboardItems
    participant DB

    Agent->>applyCommand: "FanOutDashboardItems { dashboardId, sourceItemId, field, placements[] }"
    applyCommand->>fanOutDashboardItems: validated payload

    fanOutDashboardItems->>fanOutDashboardItems: validate placements (non-empty, id/x/y types, value key present)
    fanOutDashboardItems->>DB: requireDashboardItems(dashboardId)
    DB-->>fanOutDashboardItems: DashboardItem[] (current layout)

    fanOutDashboardItems->>fanOutDashboardItems: locate source item (throws if missing or markdown)
    fanOutDashboardItems->>fanOutDashboardItems: guard duplicate ids vs existing layout

    loop for each placement p
        fanOutDashboardItems->>fanOutDashboardItems: "baseFilters = sourceFilters.filter(f != field)"
        fanOutDashboardItems->>fanOutDashboardItems: "pin = { field, op:eq, value: p.value }"
        fanOutDashboardItems->>fanOutDashboardItems: "clone = { ...source dims, overrides: { filters: [...baseFilters, pin] } }"
    end

    fanOutDashboardItems->>DB: "update layout = [...items, ...clones]"
    DB-->>fanOutDashboardItems: ok
    fanOutDashboardItems-->>applyCommand: "{ ok: true, created: string[] }"
    applyCommand-->>Agent: result
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Agent
    participant applyCommand
    participant fanOutDashboardItems
    participant DB

    Agent->>applyCommand: "FanOutDashboardItems { dashboardId, sourceItemId, field, placements[] }"
    applyCommand->>fanOutDashboardItems: validated payload

    fanOutDashboardItems->>fanOutDashboardItems: validate placements (non-empty, id/x/y types, value key present)
    fanOutDashboardItems->>DB: requireDashboardItems(dashboardId)
    DB-->>fanOutDashboardItems: DashboardItem[] (current layout)

    fanOutDashboardItems->>fanOutDashboardItems: locate source item (throws if missing or markdown)
    fanOutDashboardItems->>fanOutDashboardItems: guard duplicate ids vs existing layout

    loop for each placement p
        fanOutDashboardItems->>fanOutDashboardItems: "baseFilters = sourceFilters.filter(f != field)"
        fanOutDashboardItems->>fanOutDashboardItems: "pin = { field, op:eq, value: p.value }"
        fanOutDashboardItems->>fanOutDashboardItems: "clone = { ...source dims, overrides: { filters: [...baseFilters, pin] } }"
    end

    fanOutDashboardItems->>DB: "update layout = [...items, ...clones]"
    DB-->>fanOutDashboardItems: ok
    fanOutDashboardItems-->>applyCommand: "{ ok: true, created: string[] }"
    applyCommand-->>Agent: result
```

</a>
</details>

<sub>Reviews (3): Last reviewed commit: ["fixup! \[YW-308\] feat(dashboard): fan-out..."](https://github.com/youhaowei/dashframe/commit/adcae106bf0ba64d8cc13054536e337c08b35bbd) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=40006403)</sub>

### CODE REVIEW

File: apps/server/src/functions/commands.ts [hunk 1]
```diff
@@ -1490,6 +1504,137 @@ const removeDashboardItem = mutation({
...
+        throw new Error(
+          "FanOutDashboardItems: each placement must have a string id",
+        );
+      }
+      if (typeof p.x !== "number" || typeof p.y !== "number") {
+        throw new Error(
+          "FanOutDashboardItems: each placement must have numeric x and y",
+        );
+      }
+    }
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts:1593]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts [hunk 2]
```diff
@@ -1490,6 +1504,137 @@ const removeDashboardItem = mutation({
...
+
+    // Build N clones. Each clone inherits the source's overrides.filters (and
+    // sorts/limit) then replaces/appends the pin for `field`.
+    const sourceFilters: DashboardItemOverrides["filters"] =
+      source.overrides?.filters ?? [];
+    const sourceSorts = source.overrides?.sorts;
+    const sourceLimit = source.overrides?.limit;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts [hunk 3]
```diff
@@ -1295,6 +1295,18 @@ interface DashboardItem {
...
+/**
+ * Override bag written into `DashboardItem.overrides` by the fan-out primitive.
+ * Mirrors `DashboardItemOverrides` from @dashframe/types — kept inline to avoid
+ * a circular package dependency from the server layer.
+ */
+interface DashboardItemOverrides {
+  filters?: { field: string; operator: string; value: unknown }[];
+  sorts?: { field: string; direction: "asc" | "desc" }[];
+  limit?: number;
 }
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts:1324]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts [hunk 4]
```diff
@@ -1334,6 +1346,8 @@ function requireDashboardItem(value: unknown): DashboardItem {
       throw new Error(`AddDashboardItem: item.${key} must be a number`);
     }
   }
+  // `overrides` is passed through as-is — it is written by the fan-out primitive
+  // which controls the shape; the per-field filter pin is validated there.
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts:1364]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts:1593]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts [hunk 5]
```diff
@@ -1490,6 +1504,137 @@ const removeDashboardItem = mutation({
...
+    const sourceFilters: DashboardItemOverrides["filters"] =
+      source.overrides?.filters ?? [];
+    const sourceSorts = source.overrides?.sorts;
+    const sourceLimit = source.overrides?.limit;
+
+    const clones: DashboardItem[] = placements.map((p) => {
+      // Clone the source filters array; replace the filter for `field` in-place
+      // if one exists, otherwise append. Stable ordering for reproducibility.
+      const baseFilters = sourceFilters.filter((f) => f.field !== field);
+      const pin = { field, operator: "eq", value: p.value };
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts [hunk 6]
```diff
@@ -2008,6 +2172,7 @@ export interface DashboardItemInput {
   y: number;
   width: number;
   height: number;
+  overrides?: DashboardItemOverridesInput;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts:2225]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts [hunk 7]
```diff
@@ -1998,6 +2144,24 @@ export interface TypedInsightFilter {
   value: FilterOperandValue;
 }
 
+/**
+ * Filter override for a single dashboard item. Mirrors the domain
+ * `InsightFilterOverride` shape — kept here to avoid a shared-package coupling.
+ */
+export interface DashboardItemFilterOverride {
+  field: string;
+  operator: "eq" | "ne" | "gt" | "gte" | "lt" | "lte" | "contains" | "in";
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts]: Fixed in adcae106. See commit message for details.

File: apps/server/src/functions/commands.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/commands.ts:1364]: Fixed in adcae106. See commit message for details.
